pkg/websocket/client.go

```diff
@@ -162,8 +163,10 @@ func (c *Client) Run(ctx context.Context) {
 			c.cfg.Log.WithFields(log.Fields{
 				"prefix": "websocket.Client.Run",
 			}).Debug("Resetting the connection")
+			c.isResetting = true
 			c.Close(ws.CloseNormalClosure, "Resetting the connection")
 			c.wg.Wait()
+			c.isResetting = false
```

Comment:
I don't know if this will always be effective at preventing the deadlock. Possible that I need some proper synchronization between the go routines?

Reply:
This is probably fine - what I'm wondering is 
1. why are we not `wg.Wait`ing for the 2 first switch cases
2. should we set/unset `c.isResetting` in switch cases, just to keep things more consistent
3. depending on how much we want Run and read/writePump to be sync/aware from each other, we could also make the `notifyClose` chan buffered, which would avoid making sends to it blocking. ReadPump and WritePump both send to it, so buffer of len 2 would prob be enough.

I'm honestly not sure what the best is, we can discuss tomorrow 1:1 about the pros/cons of each.

Reply:
I need the go equivalent of a promise. I want close to only run one-at-a-time, but if it is already running it should return the current running promise if that makes sense.

Reply:
Boolean state was removed in favor of select on channels.